commons/src/main/java/org/infinispan/commons/api/functional/FunctionalMap.java

```diff
@@ -0,0 +1,634 @@
+       */
+      CompletableFuture<Void> eval(K key, V value, BiConsumer<V, WriteEntryView<V>> f);
+
+      /**
+       * Evaluate a write-only {@link Consumer} operation with a
+       * {@link WriteEntryView} of the value associated with the key,
+       * and return a {@link CompletableFuture} which will be
+       * completed with the object returned by the operation.
+       *
+       * By returning {@link CompletableFuture} instead of the function's
+       * return type directly, the method hints at the possibility that to
+       * executing the function might require remote data present in either
+       * a persistent store or a remote clustered node.
+       *
+       * DESIGN RATIONALES:
+       * <ul>
+       *    <li>Since this is a write-only operation, no entry attributes can be
+       *    queried, hence the only reasonable thing can be returned is Void.
+       *    </li>
+       *    <li>Why provide this operation? Isn't {@link #eval(Object, Object, BiConsumer)} enough?
+       *    Possibly, this operation makes a bit simpler to write constant
+       *    values along with optional metadata parameters.
+       *    </li>
+       * </ul>
+       *
+       * @param key the key associated with the {@link WriteEntryView} to be
+       *            passed to the operation
+       * @param f operation that takes a {@link WriteEntryView} associated with
+       *          the key and writes to the it without returning anything
+       * @return a {@link CompletableFuture} which will be completed the
+       *         operation completes
+       */
+      CompletableFuture<Void> eval(K key, Consumer<WriteEntryView<V>> f);
+
+      /**
+       * Evaluate a write-only {@link BiConsumer} operation, with a value
+       * passed in and a {@link WriteEntryView} of the value associated with
+       * the key, for each of the keys in the set passed in, and
+       * returns a {@link CloseableIterator} that allows navigating completion
+       * of the operation executions.
+       *
+       * This method can be used to implement operations such as:
+       *
+       * <ul>
+       *    <li>{@link ConcurrentMap#putAll(Map)}</li>
+       *    <li>{@link javax.cache.Cache#putAll(Map)}</li>
+       * </ul>
+       *
+       * DESIGN RATIONALE:
+       * <ul>
+       *    <li>It makes sense to expose global operation like this
+       *    instead of forcing users to iterate over the keys to lookup and call
+       *    get individually since Infinispan can do things more efficiently.
+       *    </li>
+       *    <li>Since this is a write-only operation, no entry attributes can be
+       *    queried, hence the only reasonable thing can be returned is Void.
+       *    </li>
+       *    <li>It still makes sense to return a CloseableIterator<Void> instead of
+       *    CompletableFuture<Void> in case the user wants to do something as each
+       *    operation completes, e.g. keep track of progress.
```

Comment:
I don't think the iterator is worth it here either. For a local cache completing each modification as it happens in the data container seems like a lot of overhead for something that's very unlikely to be used. For a distributed cache it's even more complicated, we can't advance the iterator at all until we have committed the entries in the local data container - before that there's always a chance of having to retry the operation.